sql: Add support for parsing ALTER COLUMN TYPE

bobvawter opened the pull request:
This change adds SQL parsing support for the proposed ALTER COLUMN TYPE syntax
in RFC #24703.  It still returns an unimplemented error message to the caller.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24958)
<!-- Reviewable:end -->

mjibson commented:
The BNF generators (diagrams.go in cmd/docgen) need to be updated. I can help out if needed: they are kinda annoying. For instance,  add_column.bnf needs to be unchanged since this isn't an add column.

Also add me as a reviewer next time so I see these PRs. I just found it while looking for other stuff.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/sem/tree/alter_table.go, line 164 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24958#-LAlALnIBGOVVXHTOZE3:-LAlALnIBGOVVXHTOZE4:b-bfqgmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb05c31d08cb0e5e81162b3d0a32af3ec59299a6/pkg/sql/sem/tree/alter_table.go#L164)):*
> ```Go
> 
> // Format implements the NodeFormatter interface.
> func (node *AlterTableAlterColumnType) Format(ctx *FmtCtx) {
> ```

This needs tests. See TestParse in parse_test.go.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24958#-:-LAl8cQWCq-K7vK9HF9_:b3kdqfg)*
<!-- Sent from Reviewable.io -->

bobvawter committed:
sql: Add support for parsing ALTER COLUMN TYPE

This change adds SQL parsing support for the proposed ALTER COLUMN TYPE syntax
in RFC #24703. It still returns an unimplemented error message to the caller.

Release note: None

bobvawter commented:
PTAL: Updated docgen and verified that the resulting BNF result in pretty pictures.

bobvawter commented:
bors r+
